# 디버깅

## 디버깅의 정의

- 잘못된 프로그램을 수정하는 것을 디버깅이라함
- de(없앤다)+ bugging(버그)

## 에러

### 문법에러(Syntax Error)

- 파일이름, 줄번호, ^문자를 통해 파이썬이 코드를 읽어나갈 때(parser) 문제가 발생한 위치를 표현
- 줄에서 에러가 감지된 가장 앞의 위치를 가리키는 캐럿(^)기호를 표시

- **invalid syntax** 문법 오류
- **assign to literal** 잘못된 할당
- **EOL (End Of Line)** 보통 괄호 안 닫으면 발생
- **EOF (End Of File)** 보통 괄호 안 닫으면 발생

## 예외(exception)

- 실행 도중 예상치 못한 상황을 맞이하면 프로그램 실행을 멈춤
- 문자나 표현식이 문법적으로 올바르더라도 발생하는 에러
- 실행중에 감지되는 에러들을 예외라고 부름

- **ZeroDivisionError**

```python
10/0 #ZeroDivisionError : division by zero
     #0으로 나누고자 할 때 발생
```

- **NameError**

```python
print(name_error)
#NameError : name 'name_error' is not defined
#namespase상에 이름이 없는 경우(Not defined)
```

- **TypeError**
    - 타입 불일치
    
    ```python
    1 + '1' #TypeError : unsupported operand type(s) for +: 'int' and 'str'
    round('3.5') # #TypeError : type str doesn't define __round__ method
    ```
    
    - argument 누락
    
    ```python
    divmod() #TypeError : divmod expected 2 arguments, got 0
    
    import random
    random.sample() 
    #TypeError : sample() missing 2 required positional arguments: 'population' and 'k'
    ```
    
    - argument 개수 초과
    
    ```python
    divmod(1,2,3) #TypeError : divmod expected 2 arguments, got 3
    
    import random
    random.sample(range(3),1,2) 
    #TypeError : sample() takes 3 positional arguments but 4 were given
    ```
    
    - argument 불일치
    
    ```python
    import random
    random.sample(1,2)
    # TypeError : Population must be a sequence. For dicts or sets, use sorted.
    (sequence들고와)
    ```
    
- ValueError

```python
# 타입은 올바르나 값이 적절하지 않거나 없는 경우
int('3.5') #ValueError : invalid literal for int() with base 10: 3.5
range(3).index(6) #ValueError : 6 is not in range
```

- IndexError

```python
# 인덱스가 존재하지 않거나 범위를 벗어나는 경우
empty_list = []
empty_list[2]
#IndexError : list index out of range
```

- Keyerror

```python
song = {'IU' : '좋은날'}
song['BTS']
```

- ModuleNotFoundError

```python
# 설치가 잘 안됐거나 위치를 잘못잡아서 모듈을 못찾을때
import ssafy
```

- ImportError

```python
# 모듈은 있으나 존재하지 않는 클래스/함수를 가져오는 경우
from random import sample

from random import samp
```

- KeyboardInterrupt

```python
# 임의로 키보드로 프로그램을 종료했을 때
```

- IndentedError

```python
# 들여쓰기 
```

- 파이썬의 내장 예외

## 예외 처리

- **try 문(statement)**
    - 오류가 발생할 가능성이 있는 코드를 실행
    - 예외가 발생되지 않으면, except 없이 실행 종료
- **except 문(clause)**
    - 예외가 발생하면 except 절이 실행
    - 예외 상황을 처리하는 코드를 받아서 적절한 조치를 취함
    

![Untitled](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/5dc18238-2cb7-4a59-98d6-03e416fc3c4b/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20220731%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220731T144244Z&X-Amz-Expires=86400&X-Amz-Signature=6e6df4e59666c62c1c8de0aee0f27fa5c12ade4a2c0d03c3627f77005b906789&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22&x-id=GetObject)

### 예외 처리 예시

```python
try:
		num = input('숫자 입력:')  #try 명령문
		print(int(num))           
except ValueError:            #예외그룹
		print('숫자가 입력되지 않았습니다.')  #예외 처리 명령문
(finally)
```

### 에러 메세지 처리(as)

- as키워드를 활용하여 원본 에러 메시지를 사용할 수 있음
    - 예외를 다른 이름에 대입

![Untitled](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/b09b3e55-c415-42e3-a848-b952f49f2cab/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20220731%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220731T144304Z&X-Amz-Expires=86400&X-Amz-Signature=dcc3a8060f7594f27e26baf4259eebb5c719149e31770d70cc86db6888f153e3&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22&x-id=GetObject)

### 예외  처리 종합

- try
    - 코드를 실행함
- except
    - try문에서 예외가 발생 시 실행
- else
    - try문에서 예외가 발생하지 않으면 실행
- finally
    - 예외 발생 여부와 관계없이 항상 실행함